File: vfr-green-app/app/src/main/java/com/blueskycharts/app/map/view/MapTileView.kt
package com.blueskycharts.app.map.view

import android.graphics.*
import com.blueskycharts.app.coordinates.*
import com.blueskycharts.app.map.models.SubMapModel
import com.blueskycharts.app.map.resources.TileProvider
import com.blueskycharts.app.map.resources.TileReceiver
import kotlin.math.*

class MapTileView(private val tileProvider: TileProvider, private val map: Map) : SubMapView, TileReceiver {
    // Metadata
    override var fileExtentMercator: BoxWebMercator? = null
        get() {
            val extent = this.fileExtent ?: return null
            if ( field == null ) {
                field = CoordinateConversion.convertBoxGeoToBoxMercator(extent)
            }
            return field
        }
        private set
    var fileExtent2d: Box2d? = null
        get() {
            val extent = this.fileExtent ?: return null
            if ( field == null ) {
                field = CoordinateConversion.convertBoxGeoToBox2d(extent)
            }
            return field
        }
        private set

    private var fileExtent: BoxGeo? = null
    private var tileWidth: Int = 0
    private var tileHeight: Int = 0
    private val tileDimensionPercentage: Point2d = Point2d(1.0, 1.0)
    private var mapName: String = ""
    private var mapVersion: String = "0"
    private var maxZoom: Int = 0

    // Rendering
    private var isDisposed: Boolean = false

    override fun initialize(name: String, model: SubMapModel?): BoxWebMercator? {
        if (model?.imageHeight == null || model.imageWidth == null || model.tileWidth == null ||
            model.version == null || model.fileExtent == null || model.maxZoom == null)
            return null

        this.tileWidth = model.tileWidth
        this.tileHeight = model.tileWidth
        this.mapName = name
        this.mapVersion = model.version
        this.maxZoom = model.maxZoom

        val fileExtent = model.fileExtent.createBoxGeo()
        this.fileExtent = fileExtent
        val fileExtentMercator = this.fileExtentMercator ?: return null

        if (fileExtentMercator.width > fileExtentMercator.height)
            this.tileHeight = round(this.tileWidth * fileExtentMercator.height / fileExtentMercator.width).toInt()
        else
            this.tileWidth = round(this.tileHeight * fileExtentMercator.width / fileExtentMercator.height).toInt()

        return fileExtentMercator
    }

    override fun dispose() {
        this.isDisposed = true;
    }

    override fun resetRequestedInformationAgeRecord() {
    }

    override fun getRequestedInformationAgeSeconds(): Int? {
        return null;
    }

    override fun receiveTile(
            subsection: Box2d,      // The subsection (in percentage) of the tile being provided that should be drawn.
                                    // This may be less than (0,0,1,1) if a substitute lower zoom tile is being provided
            tile: Bitmap?,          // The actual tile image to draw
            data: Any?,             // A renderData object specifying the location to draw the tile.  This must be
                                    // modified by the subsection
            immediate: Boolean,     // True if this function is being called on the main thread
            canvas: Canvas?         // The canvas to draw on
    ) {
        if ( this.isDisposed || tile == null ||
             !immediate || canvas == null ||
             data == null || data !is RenderData ) {
            this.map.requestRedraw()
            return
        }

        val restoreTo = canvas.save()

        val sourceRect = Rect(
            (subsection.upperLeft.x * tile.width).toInt(),
            (subsection.upperLeft.y * tile.height).toInt(),
            (subsection.lowerRight.x * tile.width).toInt(),
            (subsection.lowerRight.y * tile.height).toInt()
        )
        val destinationRect = RectF(
            data.destination.upperLeft.x.toFloat(),
            data.destination.upperLeft.y.toFloat(),
            data.destination.lowerRight.x.toFloat(),
            data.destination.lowerRight.y.toFloat()
        )
        canvas.drawBitmap(tile, sourceRect, destinationRect, null)

        canvas.restoreToCount(restoreTo);
    }

    /**
     *
     * @param context
     * @param region The region of the map to draw
     * @param destination The actual pixels to draw on (this must be in screen pixel coordinates so zoom level can be determined)
     */
    override fun render(canvas: Canvas, region: BoxWebMercator, destination: Box2d) {
        if (this.isDisposed) {
            return;
        }
        val fileExtentMercator = this.fileExtentMercator ?: return

        // Figure out the Box2d for the full map
        val regionPercentage = fileExtentMercator.overlapPercentageUpperLeft(region)
        val fullMapDestinationWidth = (destination.upperLeft.x - destination.lowerRight.x) / (regionPercentage.upperLeft.x - regionPercentage.lowerRight.x)
        val fullMapDestinationX = destination.upperLeft.x - fullMapDestinationWidth * regionPercentage.upperLeft.x
        val fullMapDestinationHeight = (destination.upperLeft.y - destination.lowerRight.y) / (regionPercentage.upperLeft.y - regionPercentage.lowerRight.y)
        val fullMapDestinationY = destination.upperLeft.y - fullMapDestinationHeight * regionPercentage.upperLeft.y

        // Figure out the zoom level for what we are drawing
        val zoom0TilePixelsPerMercator = this.tileWidth / fileExtentMercator.width
        val desiredPixelsPerMercator = destination.width / region.width
        var zoomLevel = ceil(log2(desiredPixelsPerMercator / zoom0TilePixelsPerMercator)).toInt()
        if (zoomLevel < 0)
            zoomLevel = 0
        if (zoomLevel > this.maxZoom)
            zoomLevel = this.maxZoom

        // Figure out what tiles we need to draw
        val numTilesAcross = 2.0.pow(zoomLevel).toInt()
        var startX = floor(regionPercentage.upperLeft.x * numTilesAcross).toInt()
        var startY = floor(regionPercentage.upperLeft.y * numTilesAcross).toInt()
        var endX = ceil(regionPercentage.lowerRight.x * numTilesAcross).toInt()
        var endY = ceil(regionPercentage.lowerRight.y * numTilesAcross).toInt()

        if ( startX < 0 ) {
            startX = 0
        }
        if ( endX > numTilesAcross ) {
            endX = numTilesAcross
        }
        if ( startY < 0 ) {
            startY = 0
        }
        if ( endY > numTilesAcross ) {
            endY = numTilesAcross
        }

        if (endX - startX > 40 || endY - startY > 40) {
            return
        }

        // Get all the images to draw
        for (x in startX until endX) {
            for (y in startY until endY ) {
                // Figure out where to draw this tile
                val tileDestination = Box2d(
                    fullMapDestinationX + fullMapDestinationWidth * (x / numTilesAcross.toDouble()),
                    fullMapDestinationY + fullMapDestinationHeight * (y / numTilesAcross.toDouble()),
                    fullMapDestinationX + fullMapDestinationWidth * ((x+1) / numTilesAcross.toDouble()),
                    fullMapDestinationY + fullMapDestinationHeight * ((y+1) / numTilesAcross.toDouble()))

                // Get the tile image and request it be drawn
                this.tileProvider.retrieveTile(
                    mapName = this.mapName,
                    mapVersion = this.mapVersion,
                    zoomLevel = zoomLevel,
                    location = Point2d(x.toDouble(), y.toDouble()),     // The tile "id" to draw
                    tileDimensions = this.tileDimensionPercentage,      // Always 1,1.  This is the percentage of the tile to draw to the screen.
                                                                        // Modified by fileTile in the tile provider if the tile is not available but a lower zoom is.
                    receiver = this,
                    canvas = canvas,
                    data = RenderData(tileDestination)    // Describes the place to draw the full tile, not a portion of the tile
                )
            }
        }
    }

    override fun moveOffscreen() {
    }

    private data class RenderData(val destination: Box2d)
}

File: vfr-green-map/src/view/map-tile-view.ts
import { BoxGeoModel } from '../models/box-geo-model'
import { Box2d, BoxWebMercator, CoordinateConversion, Point2d } from 'coordinates'
import { ISubMapView } from './i-sub-map-view'
import { SubMapModel } from '../models/sub-map-model'
import { TileProvider } from '../resources/tile-provider'
import { ITileReceiver } from '../resources/i-tile-receiver';
import { IMap } from './i-map'

export class MapTileView implements ISubMapView, ITileReceiver {
    // Metadata
    private map: IMap;
    private originalMapWidth: number;
    private originalMapHeight: number;
    private imageWidthScale: number;
    private imageHeightScale: number;
    private tileWidth: number;
    private tileHeight: number;
    private tileDimensionPercentage: Point2d;
    private mapName: string;
    private mapVersion: string;
    private maxZoom: number;

    // Rendering
    private tileProvider: TileProvider;
    private renderVersion: number;
    private context: CanvasRenderingContext2D | null;
    private contextTransform: DOMMatrix | null;
    private scaledTileWidth: number;
    private isDisposed: boolean;
    
    constructor(tileProvider: TileProvider, map: IMap) {
        this.map = map;
        this.originalMapWidth = 0;
        this.originalMapHeight = 0;
        this.imageWidthScale = 1;
        this.imageHeightScale = 1;
        this.tileWidth = 0;
        this.tileHeight = 0;
        this.tileDimensionPercentage = new Point2d();
        this.mapName = "";
        this.mapVersion = "0";
        this.maxZoom = 0;
        this.tileProvider = tileProvider;
        this.renderVersion = 0;
        this.context = null;
        this.contextTransform = null;
        this.scaledTileWidth = 0;
        this.isDisposed = false;
    }

    public initialize(model: SubMapModel, name: string): BoxWebMercator | undefined {
        if (model.imageHeight == null || model.imageWidth == null || model.tileWidth == null || 
            model.version == null || model.fileExtent == null || model.maxZoom == null)
            return undefined;

        if ( model.imageWidthScale !== undefined ) {
            this.imageWidthScale = model.imageWidthScale;
        }
        if ( model.imageHeightScale !== undefined ) {
            this.imageHeightScale = model.imageHeightScale;
        }

        this.originalMapWidth = model.imageWidth;
        this.originalMapHeight = model.imageHeight;
        this.tileWidth = model.tileWidth;
        this.tileHeight = model.tileWidth;
        this.tileDimensionPercentage = new Point2d(1, 1);
        this.mapName = name;
        this.mapVersion = model.version;
        this.maxZoom = model.maxZoom;

        if (this.originalMapWidth > this.originalMapHeight)
            this.tileHeight = Math.round(this.tileWidth * this.originalMapHeight / this.originalMapWidth)
        else
            this.tileWidth = Math.round(this.tileHeight * this.originalMapWidth / this.originalMapHeight)

        let fileExtent = BoxGeoModel.createBoxGeoFromModel(model.fileExtent);
        let fileExtent2dUpperLeft = CoordinateConversion.convertToWebMercator(fileExtent.getTopLeft());
        let fileExtent2dLowerRight = CoordinateConversion.convertToWebMercator(fileExtent.getBottomRight());
        let fileExtent2d = new BoxWebMercator(fileExtent2dUpperLeft.x, fileExtent2dUpperLeft.y, fileExtent2dLowerRight.x, fileExtent2dLowerRight.y);
        return fileExtent2d
    }

    public dispose(): void {
        this.isDisposed = true;
    }

    public getOriginalWidth(): number {
        return this.originalMapWidth;
    }

    public getOriginalHeight(): number {
        return this.originalMapHeight;
    }

    public resetRequestedInformationAgeRecord(): void {
    }

    public getRequestedInformationAgeSeconds(): number | undefined {
        return undefined;
    }

    public receiveTile(location: Point2d, subsection: Box2d, tile: HTMLImageElement, data: any, immediate: boolean): void {
        if ( this.isDisposed || data as number != this.renderVersion )
            return;

        if ( !immediate ) {
            this.map.requestRedraw();
            return;
        }

        if ( this.contextTransform == null || this.context == null )
            return;

        let tileX = location.x * this.scaledTileWidth * this.imageWidthScale;
        let tileWidth = this.scaledTileWidth * this.imageWidthScale;
        let scaledTileHeight = this.scaledTileWidth * this.tileHeight / this.tileWidth
        let tileY = location.y * scaledTileHeight * this.imageHeightScale;
        let tileHeight = scaledTileHeight * this.imageHeightScale;
        
        let currentTransform = this.context.getTransform();
        this.context.setTransform(this.contextTransform);
        let upperLeft = subsection.getUpperLeft();
        let dimensions = subsection.getDimensions();
        this.context.drawImage(tile, upperLeft.x * tile.width, upperLeft.y * tile.height, dimensions.x * tile.width, dimensions.y * tile.height,
            tileX, tileY, tileWidth, tileHeight);
        this.context.setTransform(currentTransform);
    }

    /**
     * 
     * @param context 
     * @param region The region of the map to draw relative to the original size of the map image
     * @param scale The scale to draw the map at.  Point (0,0) is the center of the map.
     */
    public render(context: CanvasRenderingContext2D, region: Box2d, scale: number) {
        if (this.isDisposed) {
            return;
        }

        // Validate the region
        region = region.clone();
        if ( region.getUpperLeft().x < 0 )
            region.getUpperLeft().x = 0;
        else if ( region.getUpperLeft().x >= this.originalMapWidth )
            region.getUpperLeft().x = this.originalMapWidth;
        if ( region.getUpperLeft().y < 0 )
            region.getUpperLeft().y = 0;
        else if ( region.getUpperLeft().y >= this.originalMapHeight )
            region.getUpperLeft().y = this.originalMapHeight;

        if ( region.getLowerRight().x < 0 )
            region.getLowerRight().x = 0;
        else if ( region.getLowerRight().x >= this.originalMapWidth )
            region.getLowerRight().x = this.originalMapWidth;
        if ( region.getLowerRight().y < 0 )
            region.getLowerRight().y = 0;
        else if ( region.getLowerRight().y >= this.originalMapHeight )
            region.getLowerRight().y = this.originalMapHeight;

        // Increase the render version to prevent old requests from rendering
        this.renderVersion++
        this.context = context;
        this.contextTransform = context.getTransform();

        // Figure out the size of what we are drawing
        let m = this.originalMapWidth * scale
        let zoomLevel = Math.ceil(Math.log(m / this.tileWidth) / Math.log(2))
        if (zoomLevel < 0)
            zoomLevel = 0;
        if (zoomLevel > this.maxZoom)
            zoomLevel = this.maxZoom;
        let numTilesAcross = 2 ** zoomLevel;
        this.scaledTileWidth = m / numTilesAcross;
        let originalImageTileWidth = this.originalMapWidth / numTilesAcross
        let originalImageTileHeight = this.originalMapHeight / numTilesAcross

        // Ensure we aren't looping too much
        let startX = Math.floor(region.getUpperLeft().x / originalImageTileWidth);
        let endX = Math.ceil(region.getLowerRight().x / originalImageTileWidth);
        let startY = Math.floor(region.getUpperLeft().y / originalImageTileHeight);
        let endY = Math.ceil(region.getLowerRight().y / originalImageTileHeight)

        if (endX - startX > 20 || endY - startY > 20)
            return;

        // Get all the images to draw
        for (let x = startX; x < endX; x++ ) {
            for (let y = startY; y < endY; y++ ) {
                // Get the tile image and request it be drawn
                this.tileProvider.retrieveTile(this.mapName, this.mapVersion, zoomLevel, new Point2d(x,y), 
                    this.tileDimensionPercentage, this, this.renderVersion);
            }
        }
    }

    public moveOffscreen(): void {
        this.context = null;
    }
}
